runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2850,7 +2882,51 @@ J9::Z::TreeEvaluator::inlineStringHashCode(TR::Node* node, TR::CodeGenerator* cg
    cg->stopUsingRegister(registerVB);
    cg->stopUsingRegister(registerVC);
 
-   return node->setRegister(registerHash);
+   node->setRegister(registerHash);
+
+   // Cleanup nodes before returning
+   cg->decReferenceCount(nodeValue);
+   cg->decReferenceCount(nodeIndex);
+   cg->decReferenceCount(nodeCount);
+   if (nodeHash)
+      cg->decReferenceCount(nodeHash);
+
+   return registerHash;
+   }
+
+TR::Register*
+J9::Z::TreeEvaluator::inlineStringHashCode(TR::Node* node, TR::CodeGenerator* cg, bool isCompressed)
+   {
+   return hashCodeHelper(node, cg, isCompressed ? TR::Int8 : TR::Int16, NULL, false);
+   }
+
+TR::Register*
+J9::Z::TreeEvaluator::inlineVectorizedHashCode(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::Register* registerHash = NULL;
+
+   switch (node->getChild(4)->getConstValue())
+      {
+      // The following constants come from the values for the type operand of the NEWARRAY instruction
+      // See https://docs.oracle.com/javase/specs/jvms/se9/html/jvms-6.html#jvms-6.5.newarray.
+      case 4:  // T_BOOLEAN
+         registerHash = hashCodeHelper(node, cg, TR::Int8, node->getChild(3), false);
+         break;
+      case 8:  // T_BYTE
+         registerHash = hashCodeHelper(node, cg, TR::Int8, node->getChild(3), true);
+         break;
+      case 5:  // T_CHAR
```

Comment:
Looking at the implementation on JDK [1], I see that in case of T_CHAR, it checks if the array is instance of byte array in which case it calls UTF16HashCode. I am just starting review, so probably will find answer myself, but just posting here, we do ensure that functionality right?

[1]. https://github.ibm.com/runtimes/openj9-openjdk-jdk21-zos/blob/a433352b229e3efbbc0689edccd6f3ef7ee68e2b/src/java.base/share/classes/jdk/internal/util/ArraysSupport.java#L204-L206

Reply:
`utf16hashCode` requires that its length and index arguments be scaled to char array length and index at the call site of  `vectorizedHashCode` (see comment at definition [1] and usage in StringUTF16 [2]). Since the scaling is done by the caller, there is nothing we need to do in the acceleration to ensure this functionality. I've included this case in my unit testing and haven't seen any problems.

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/48ccb42697017bcd03c2ed2b346d43949152308b/src/java.base/share/classes/jdk/internal/util/ArraysSupport.java#L256
[2] https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/48ccb42697017bcd03c2ed2b346d43949152308b/src/java.base/share/classes/java/lang/StringUTF16.java#L418